tutorial/lesson_17_predicated_rdom.cpp

```diff
@@ -0,0 +1,233 @@
+        // this, we need to define three predicates, where each corresponds to
+        // one side of the triangle.
+        Func f("triangle");
+        Var x("x"), y("y");
+        f(x, y) = x + y;
+        // First, let's define the minimal bounding box over the triangular
+        // region.
+        RDom r(0, 8, 0, 10, "r");
+        // Next, let's add the first predicate to the RDom using the RDom::where
+        // directive.
+        r.where(r.x + r.y > 5);
+        // Add the 2nd and the 3rd predicates.
+        r.where(3*r.y - 2*r.x < 15);
+        r.where(4*r.x - r.y < 20);
+
+        // We can also pack the multiple predicates into one like so:
+        // r.where((r.x + r.y > 5) && (3*r.y - 2*r.x < 15) && (4*r.x - r.y < 20));
+
+        // Then define the update.
+        f(r.x, r.y) *= 2;
+
+        Buffer<int> halide_result = f.realize(10, 10);
+
+        // See figures/lesson_17_rdom_triangular.gif for a visualization of
+        // what this did.
+
+        // The equivalent C is:
+        int c_result[10][10];
+        for (int y = 0; y < 10; y++) {
+            for (int x = 0; x < 10; x++) {
+                c_result[y][x] = x + y;
+            }
+        }
+        for (int r_y = 0; r_y < 10; r_y++) {
+            for (int r_x = 0; r_x < 8; r_x++) {
+                // Update is only performed if the predicate evaluates to true.
+                if ((r_x + r_y > 5) && (3*r_y - 2*r_x < 15) && (4*r_x - r_y < 20)) {
+                    c_result[r_y][r_x] *= 2;
+                }
+            }
+        }
+
+        // Check the results match:
+        for (int y = 0; y < 10; y++) {
+            for (int x = 0; x < 10; x++) {
+                if (halide_result(x, y) != c_result[y][x]) {
+                    printf("halide_result(%d, %d) = %d instead of %d\n",
+                           x, y, halide_result(x, y), c_result[y][x]);
+                    return -1;
+                }
+            }
+        }
+    }
+
+    {
+        // The predicate is not limited to RDom's variables only (e.g. r.x, r.y, etc.).
+        // It could refer to free variables in the update definition or calls to
+        // other Funcs, or make recursive calls to the same Func. Here is some
+        // example:
+        Func f("call_f"), g("call_g");
+        Var x("x"), y("y");
+        f(x, y) = 2 * x + y;
+        g(x, y) = x + y;
+
+        // This RDom's predicates depend on the initial value of 'f'.
+        RDom r1(0, 5, 0, 5, "r1");
+        r1.where(f(r1.x, r1.y) >= 10);
+        r1.where(f(r1.x, r1.y) != 15);
+        f(r1.x, r1.y) += 1;
+        f.compute_root();
+
+        // While this one involves calls to another Func.
+        RDom r2(1, 3, 1, 3, "r2");
+        r2.where(f(r2.x, r2.y) < 15);
+        g(r2.x, r2.y) += f(r2.x, r2.y)/2;
+
+        Buffer<int> halide_result_f = f.realize(5, 5);
+        Buffer<int> halide_result_g = g.realize(5, 5);
+
+        // See figures/lesson_17_rdom_calls_in_predicate.gif for a visualization
```

Comment:
This should be two gifs - one per realize call.

In the left half, what's with the blue flash? In the right half, why don't I see the loads from f done by g? I suspect something got messed up by dumping both traces into one gif.